File: javascript/rpn_eval.js
/* expression in infix notation
      (((4 + 3) * 3) + ((2 + 2) * (3 + 8)))
*/
var expr = '4 3 + 3 * 2 2 + 3 8 + * +';
var stack = [];

for (var i = 0; i < expr.length; i++) {
    if (expr[i] != ' ' && isFinite(expr[i])) {
        stack.push(Number(expr[i]));
    }
    else if (expr[i] == '+' || expr[i] == '*'){
        var ans;
        if (expr[i] == '+') {
	    ans = 0;
	}
	else if (expr[i] == '*') {
	    ans = 1;
	}
        for (j = 1 ; j <= 2; j++) {
	    if (expr[i] == '+') {
		ans += Number(stack.pop());
	    }
	    else if (expr[i] == '*') {
		ans *= Number(stack.pop());
	    }
	}
	stack.push(ans);
    }
}

console.log('Ans:', stack[0]);

File: algorithms_2/rpn_eval.py
operators = ['+', '*', '-', '/']

def rpn_eval(expr):
  stack = []
  for token in expr:
    if token not in operators:
      stack.append(token)
    else:
      if token == '+':
        ans = stack.pop()
        ans += stack.pop()
      elif token == '*':
        ans = stack.pop()
        ans *= stack.pop()
      elif token == '-':
        ans = stack[len(stack) - 2]
        ans -= stack.pop()
        stack.pop()
      elif token == '/':
        ans = stack[len(stack) - 2]
        ans /= stack.pop()
        stack.pop()
      stack.append(ans)

  return stack[0]

print(rpn_eval([5, 10, '/']))
